study1/week6/NumDecodings.java

```diff
@@ -0,0 +1,33 @@
+class NumDecodings {
+    public int numDecodings(String s) {
+
+        int n = s.length();
+        int[] dp = new int[n + 1];
+        dp[0] = 1;
+        dp[1] = s.charAt(0) != '0' ? 1 : 0;
+
+        for (int i = 2; i <= n; i++) {
+
+            int first = Integer.valueOf(s.substring(i-1, i));
```

Comment:
Converting string into `Integer` is a bit more operation than it's needed. 
For line 13 you can do `s.charAt(i-1) != '0'`
For line 16 you can do `if (s.charAt(i-2) == '1' ||  s.charAt(i-2) == '2' && s.charAt(i-1) <= '6')`